hail/python/test/hailtop/batch/test_batch.py

```diff
@@ -88,6 +88,26 @@ def test_single_job(self):
 
             assert self.read(output_file.name) == msg
 
+    def test_single_job_with_shell(self):
+        with tempfile.NamedTemporaryFile('w') as output_file:
+            msg = 'hello world'
+
+            b = self.batch()
+            j = b.new_job(shell='/bin/sh')
+            j.command(f'echo "{msg}" > {j.ofile}')
+
+            b.write_output(j.ofile, output_file.name)
+            b.run()
+
+            assert self.read(output_file.name) == msg
+
+    def test_single_job_with_nonsense_shell(self):
+        b = self.batch()
+        j = b.new_job(shell='/bin/ajdsfoijasidojf')
+        j.image('google/cloud-sdk:237.0.0-alpine')
```

Comment:
Why are you overriding the image?  If you need an image, you should use `ubuntu:18.04` or something based on it.  Since this is used (nearly) everywhere, it is the fastest because it is always cached.

Parameterizing this is on the TODO (either when we upgrade to 20.04 or the Google managed images).  When that happens, we'll search everywhere for `ubuntu:18.04` to do the replacement, and this would get missed.

Reply:
Thanks for the explanation, will change (I had it as the google image because that was the default_shell for the BatchTests).

> Why are you overriding the image? 
To trigger the branch of LocalBackend._run that previously specified `/bin/bash`. Without job._image specified, commands are executed directly in a system shell, using `sp.check_call(script, shell=True)`, which defaults to `/bin/sh` on POSIX systems (COMSPEC env variable on Windows). This can be overridden using the `executable` argument, but I didn't want to introduce complexity:

With job._image specified, the docker command is also executed using `sp.check_call(script, shell=True)`, so if we want to use `sp.check_call(script, shell=True, executable=job._shell or job._default_shell)` we'll need to decide whether we check_call's shell to be overridden when job._image is specified. Seemed to be a bit out of scope, but happy to address now, you all probably have a plan here.